### PR TITLE
Sockets - Carla

### PR DESCRIPTION
**Pull request by carlabosco**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I used a sort_media method used for the media lists.
Describe how you approached testing that model method. What edge cases did you come up with? | I still need to add this test. Sorry!
What are `session` and `flash`? What is the difference between them? | Session persists until the user closes the browser. Flash lasts until the following request cycle.
What was one thing that you gained more clarity on through this assignment? | Having control of all aspects of the project from start to finish was extremely helpful in understanding the MVC pattern.
What is the Heroku URL of your deployed application? | https://project-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I had a lot of issues when trying to deploy to heroku that I don't quite understand. Maybe some heroku troubleshooting insights could be helpful.

### CODE REVIEW

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    works.sort_by! { |work| Vote.where(work_id: work.id).length }
+    unless works.nil?
+      return works.reverse[0..9]
+    end
+  end
+
+  def self.sort_media(category)
+    works = Work.where(category: category).to_a
+    works.sort_by! { |work| Vote.where(work_id: work.id).count }
+    return works.reverse
```
Comment (droberts-sea) [inline app/models/work.rb:16]: There is a lot of repeated work between this method and `top_ten`. Could you DRY this up somehow?

File: app/views/homepages/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    </section>
+
+    <hr class="root__hr">
+    <section class="top-ten__container">
+      <section class="top-ten__list_container">
+        <h2 class="top-ten__header">
+        Top Movies
+        </h2>
+        <ul class="list-group top-ten__list">
+          <% Work.top_ten("movie").each do |movie| %>
```
Comment (droberts-sea) [inline app/views/homepages/index.html.erb:22]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+module ApplicationHelper
+  def bootstrap_class_for(flash_type)
+    case flash_type
+      when "success"
+        "alert-success"
```
Comment (droberts-sea) [inline app/helpers/application_helper.rb:5]: Nice, this is a good use of a view helper.

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  def upvote
+    @current_user = User.find_by(id: session[:user_id])
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
+
+    if @current_user.nil?
+      flash[:error] = "You must be logged in to see this page"
+      redirect_to login_path
+    elsif Vote.exists?(user_id: @current_user.id, work_id: work_id)
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:85]: Would it be possible to extract the tricky logic on line 85 to the model, possibly in the form of a validation?

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    flash[:status] = :success
+    flash[:message] = "Successfully deleted book #{work.category} #{work.id}"
+    redirect_to works_path
+  end
+
+  def upvote
+    @current_user = User.find_by(id: session[:user_id])
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:80]: You have defined both `votes#create` and `works#upvote`. Probably your application only needs one of these - I think `works#upvote` is the one in use.

File: config/routes.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+Rails.application.routes.draw do
+  root 'homepages#index'
+  resources :works
+  resources :homepages
+  resources :users
```
Comment (droberts-sea) [inline config/routes.rb:5]: I don't think you need all 7 RESTful routes for `homepages` and `users` for this app.

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "Log Out", logout_path, method: :post, :class=> "btn btn-primary"  %>
+          </li>
+        <% end %>
+
+      </ul>
+    </nav>
+  </header>
+
+  <body>
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:54]: Everything that renders on the page should be inside the `<body>` tag.

File: app/views/works/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,80 @@
+<main>
+  <h2>List of Works</h2>
+  <section class="media-table">
+    <h4>Albums</h4>
+    <table class="table">
```
Comment (droberts-sea) [inline app/views/works/index.html.erb:5]: Again, this page could stand to be much DRYer.

File: test/controllers/homepages_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepagesController do
+  # it "must be a real test" do
+  #   flunk "Need real tests"
+  # end
```
Comment (droberts-sea) [inline test/controllers/homepages_controller_test.rb:6]: Looks like this was missed. In addition to the nominal case, there's an interesting edge case here when there are no works - how does the media spotlight section behave?

File: test/controllers/users_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+require "test_helper"
+
+describe UsersController do
+  describe "login" do
+    it "can log in a user" do
+
```
Comment (droberts-sea) [inline test/controllers/users_controller_test.rb:6]: There are 3 different behaviors I would be interested in testing here:
- Log in a new user
- Log in a returning user
- Fail to log in (e.g. send "" as the username)

File: test/controllers/votes_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,71 @@
+require "test_helper"
+
+describe VotesController do
+  describe "create" do
+    it "is possible to vote if logged in" do
+      user = perform_login(User.first)
```
Comment (droberts-sea) [inline test/controllers/votes_controller_test.rb:6]: Good work on the tests for this - the logic here is rather complex.

Organizationally, these target the `upvote_work_path`, which is under the `WorksController`, so I would probably put these tests in that test file.

File: test/models/work_test.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+      expect(result).must_equal false
+      expect(@work.errors.messages).must_include :title
+    end
+  end
+
+  describe 'custom methods' do
+    it 'sorts media by number of votes' do
+      Vote.all = {
+        one: {
```
Comment (droberts-sea) [inline test/models/work_test.rb:29]: There's a lot of interesting test cases for your custom `Work` methods missing here, particularly for the `top_ten` method. For example, I would ask:

- What if there are no works of that category?
- What if there are less than 10 works?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes generated
Views are well-organized (DRY, use of semantic HTML, use of partials) | some - see inline
Errors are reported to the user | yes
Business logic lives in the models | mostly - see inline
Models are thoroughly tested, including relations, validations and any custom logic | mostly - see inline
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work | mostly - see inline
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | no
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** |Good job overall. Your implementation matches the demo site pretty closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
